### PR TITLE
[#861] Implement HiDPI detection and support

### LINKED ISSUE #861: High DPI Support for Excalibur
### Steps to Reproduce
On high DPI screens, assets in Excalibur do not scale correctly as expected for the screen. In order to reproduce this run https://sarasflare.com/play/#/login 

An explanation of the issue is located here in the excalibur google group https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/excaliburjs/GgOyA8dQd8E 

We should look at investing in a high DPI screen or any emulation tools that can repro this locally.

### Expected Result
Assets should be scaled to the appropriate size on high DPI screens
image

### Actual Result
Assets are scaled incorrectly 

image

### PR DESCRIPTION
**Pull request by eonarheim**
Closes #861 

## Changes:

- Automatically detect HiDPI screens and automatically scale backing canvas to support the increase pixels
- Add option to suppress HiDPI detection for games that don't want to use it, it theoretically will cause lower FPS because the browser is driving `devicePixelRatio` more pixels
- Add convenience getters for common canvas calculations `canvasWidth`, `canvasHeight`, `halfCanvasWidth, `halfCanvasHeight`, `drawWidth`, `drawHeight`, `halfDrawWidth`, `halfDrawheight`
- Deprecated - rename getDraw to be consistent getters not methods
- Some re-organization for code locality and readibility
- Fix typos

### CODE REVIEW

**Comment by eonarheim** [comment]
@jedeen @kamranayub Please review very carefully.

@kamranayub I'd like you to test on you HiDPI machine as well

File: sandbox/tests/collision/touching.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@ var game = new ex.Engine({
     height: 200,
     width: 800
 });
-var paddle = new ex.Actor(150, game.getDrawHeight() - 40, 200, 20);
```
Comment (jedeen) [inline sandbox/tests/collision/touching.ts]: I believe our policy is to deprecate old methods and then replace them after one milestone. Could you leave these in with `@deprecated` and then move the changelog entry to 'Deprecated'?

File: sandbox/tests/collision/touching.ts [hunk 2]
```diff
@@ -23,7 +23,7 @@ ball.on('update', function () {
     if (this.pos.x < (this.getWidth() / 2)) {
         this.vel.x *= -1;
     }
-    if (this.pos.x + (this.getWidth() / 2) > game.getDrawWidth()) {
```
Comment (jedeen) [inline sandbox/tests/collision/touching.ts]: Could you leave these in with `@deprecated` and then move the changelog entry to 'Deprecated'?

File: src/engine/Engine.ts [hunk 3]
```diff
@@ -932,12 +954,117 @@ O|===|* >________________>\n\
...
+         this.canvas.height = oldHeight * this.pixelRatio;
+
+         this.canvas.style.width = oldWidth + 'px';
+         this.canvas.style.height = oldHeight + 'px';
+
+         this._logger.warn(`Hi DPI screen detected, resetting canvas resolution from 
+                           ${oldWidth}x${oldHeight} to ${this.canvas.width}x${this.canvas.height} 
+                           css size will remain ${oldWidth}x${oldHeight}`);
+         
+         let scaledWidth = this.canvasWidth / this.pixelRatio;
```
Comment (kamranayub) [inline src/engine/Engine.ts]: This is just `oldWidth` and `oldHeight`, is that right?

**Comment by eonarheim** [comment]
@kamranayub @jedeen  After much wailing and gnashing of teeth I think I've got this, please test again

**Comment by eonarheim** [comment]
@kamranayub Please review again

**Comment by kamranayub** [comment]
Works like a charm! Tested sandbox and visual tests. 💯

**Comment by eonarheim** [comment]
@kamranayub no new shrinkwrap is generated

**Comment by kamranayub** [comment]
You're confirmed using npm 5? (npm -v)

**Comment by eonarheim** [comment]
@kamranayub yup

**Comment by kamranayub** [comment]
Let me check why my copy thinks it needs an update

**Pull request merged**